File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0 , len = INT_MIN , n = s.length();
        unordered_map<char , int> mp;
        for(int r = 0 ; r < n ; r++){
            if(mp.count(s[r]) == 0 || mp[s[r]] < l){
                mp[s[r]] = r;
                len = max(len,r-l+1);
            }else{
               l = mp[s[r]] + 1;
               mp[s[r]] = r;
            }
        }
        return len == INT_MIN ? 0 : len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        res = 0
        mp = defaultdict(int)
        l , r = 0 , 0
        while r < n:
            mp[s[r]] += 1
            while mp[s[r]] > 1:
                mp[s[l]] -= 1
                l += 1
            res = max(res , r - l + 1)
            r += 1
            
        return res